main/resolve/src/mill/resolve/ResolveCore.scala

```diff
@@ -240,29 +242,39 @@ private object ResolveCore {
   }
 
   def resolveTransitiveChildren(
-      rootModule: BaseModule,
-      cls: Class[_],
-      nameOpt: Option[String],
-      segments: Segments,
-      typePattern: Seq[String]
+    rootModule: BaseModule,
+    cls: Class[_],
+    nameOpt: Option[String],
+    segments: Segments,
+    typePattern: Seq[String],
+    seenModules: Set[Class[_]],
   ): Either[String, Set[Resolved]] = {
-    val direct =
-      resolveDirectChildren(rootModule, cls, nameOpt, segments, typePattern)
-    direct.flatMap { direct =>
-      for {
-        directTraverse <-
-          resolveDirectChildren(rootModule, cls, nameOpt, segments, Nil)
-        indirect0 = directTraverse
-          .collect { case m: Resolved.Module =>
-            resolveTransitiveChildren(
-              rootModule,
-              m.cls,
-              nameOpt,
-              m.segments,
-              typePattern
-            )
+    val errOrDirect = resolveDirectChildren(rootModule, cls, nameOpt, segments, typePattern)
+    val directTraverse = resolveDirectChildren(rootModule, cls, nameOpt, segments, Nil)
+
+    val errOrModules = directTraverse.map { modules =>
+      modules.flatMap {
+        case m: Resolved.Module => Some(m)
+        case _ => None
+      }
+    }
+
+    if (seenModules.contains(cls)) {
+      Left(s"Cyclic module reference detected: ${cls.getName}, it's required to wrap it in ModuleRef. See documentation: https://mill-build.org/mill/0.12.1/fundamentals/modules.html#_abstract_modules_references")
```

Comment:
It would be better if this link used the `MILL_VERSION` or I could not give the link. WDYT?

Reply:
Ideally, the full link should be injected and generated in `build.sc`, since page refactoring happens now and then, and we could feed it to our link-checker when generating the docs.

Reply:
I like @lefou's idea but I'll consider that out of scope. It would be nice if some warnings linked to docs, but I guess that would be part of a separate effort.